programs/whirlpool/src/util/v2/token.rs

```diff
@@ -0,0 +1,378 @@
+    let extension = token_account_unpacked.get_extension::<extension::memo_transfer::MemoTransfer>();
+
+    if let Ok(memo_transfer) = extension {
+        return Ok(memo_transfer.require_incoming_transfer_memos.into());
+    } else {
+        return Ok(false);
+    }
+}
+
+pub fn is_supported_token_mint<'info>(
+    token_mint: &InterfaceAccount<'info, Mint>,
+    is_token_badge_initialized: bool,
+) -> Result<bool> {
+    let token_mint_info = token_mint.to_account_info();
+
+    // if mint is owned by Token Program, it is supported (compatible to initialize_pool / initialize_reward)
+    if *token_mint_info.owner == Token::id() {
+        return Ok(true);
+    }
+
+    // now mint is owned by Token-2022 Program
+
+    // reject native mint of Token-2022 Program to avoid SOL liquidity fragmentation
+    if spl_token_2022::native_mint::check_id(&token_mint.key()) {
+        return Ok(false);
+    }
+
+    // reject if mint has freeze_authority
+    if token_mint.freeze_authority.is_some() && !is_token_badge_initialized {
+        return Ok(false);
+    }
+
+    let token_mint_data = token_mint_info.try_borrow_data()?;
+    let token_mint_unpacked = StateWithExtensions::<spl_token_2022::state::Mint>::unpack(&token_mint_data)?;
+
+    let extensions = token_mint_unpacked.get_extension_types()?;
+    for extension in extensions {
+        match extension {
+            // supported
+            extension::ExtensionType::TransferFeeConfig => {}
+            extension::ExtensionType::TokenMetadata => {}
+            extension::ExtensionType::MetadataPointer => {}
+            // partially supported
+            extension::ExtensionType::ConfidentialTransferMint => {
+                // Supported, but non-confidential transfer only
+                //
+                // WhirlpoolProgram invokes TransferChecked instruction and it supports non-confidential transfer only.
+                //
+                // Because the vault accounts are not configured to support confidential transfer,
+                // it is impossible to send tokens directly to the vault accounts confidentially.
+                // Note: Only the owner (Whirlpool account) can call ConfidentialTransferInstruction::ConfigureAccount.
+            }
+            // supported if token badge is initialized
+            extension::ExtensionType::PermanentDelegate => {
+                if !is_token_badge_initialized { return Ok(false); }
+            }
+            extension::ExtensionType::TransferHook => {
+                if !is_token_badge_initialized { return Ok(false); }
+            }
+            extension::ExtensionType::MintCloseAuthority => {
+                if !is_token_badge_initialized { return Ok(false); }
+            }
+            extension::ExtensionType::DefaultAccountState => {
+                if !is_token_badge_initialized { return Ok(false); }
+
+                // reject if default state is not Initialized even if it has token badge
```

Comment:
What about if the default account state is frozen? I think this will mess with the token vaults of the pool. i.e. the freeze authority will need to unfreeze the vaults before the pool becomes usable

Reply:
I agree to reject mint with "Frozen" default state even if it has TokenBadge.

If the current default state is not "Initialized" (not "Initialized" = "Frozen"), I believe the current implementation rejects it by the following code (L:269-273).  Am I missing something ?

Also it is tested in the following file.
sdk/tests/integration/v2/initialize_pool_v2.test.ts
L:1076
L:1088
(Direct link doesn't work due to many diff ?)

Reply:
Got it. The `DefaultAccountState` is just en enum. I thought it was more sophisticated